### ISSUE TITLE
Inconsistent levels of severity for `cabal check`'s treatment of -O0 and -O2

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/cabal#4551]
[A Travis build](https://travis-ci.org/ku-fpg/remote-monad/jobs/239332892#L1287) broke for me recently after upgrading to `cabal-install-2.0` because `-O0` was specified in the `.cabal` file:

```
$ /opt/cabal/2.0/bin/cabal check
These warnings may cause trouble when distributing the package:
* 'ghc-options: -O0' is not needed. Use the --disable-optimization configure
flag.

$ echo $?
1
```

But erroring on `-O0` is strange, because `cabal check` does _not_ do this for `-O2`:

```
$ /opt/cabal/2.0/bin/cabal check
These warnings may cause trouble when distributing the package:
* 'ghc-options: -O2' is rarely needed. Check that it is giving a real benefit
and not just imposing longer compile times on your users.

$ echo $?
0
```

This discrepancy is strange. `cabal check` should either make both of these warnings, or both of these errors (I'd prefer the former, personally, to avoid breakage).

**Comment by hvr** [comment]
IMO, the gap between `-O0` vs `-O1`  is *much* bigger than the gap between `-O1` and `-O2` in terms of `.o` size and performance. But I have no numbers right now to back this up right now. But assuming this is right, I can see why `-O0` is treated harsher than `-O2` (which often may just waste compilation time & memory for little gain, but is otherwise harmless, except for conflicting with cabal's optimization flag handling)

**Comment by RyanGlScott** [comment]
Maybe I should ask a different question then: how can I enforce the use of `-O0` in a `.cabal` file while keeping `cabal check` happy? I can't specify `--disable-optimization` in a `.cabal` file, to my knowledge.

**Comment by 23Skidoo** [comment]
I think you can hack around this with `OPTIONS_GHC`.

**Comment by 23Skidoo** [comment]
FTR, the `-O2` check also used to cause abnormal exit, but that was changed in 0f8c45b1310b8d8f1dbf9a14e82c1ed43be507c4.

**Comment by goldfirere** [comment]
I believe @RyanGlScott and I have the same use-case here. The issue is in building the testsuite for a package (units). I do not want the testsuite optimized, because most of the tests are compile-time only. I thus want the *compile time* optimized, not the runtime. `-O0` means precisely this. :)

I *could* use `OPTIONS_GHC` in all the files in the testsuite, but that's annoying.

**Comment by hvr** [comment]
@goldfirere ok, but this sounds like a workspace desiderata rather than a package property. I.e. a local configuration, and as such it sounds like something that belongs into your workspace/project configuration, IOW your `cabal.project` file (iirc we have an open feature request somewhere), than the package meta-data that gets published & distributed on Hackage (this is more or less what `cabal check` is all about  - linting your package for distribution). So this sounds rather like a feature request to augment `cabal.project`'s syntax to allow to allow optimisation flags set on a more granular per-component level (rather than the current per-package granularity).

**Comment by 23Skidoo** [comment]
I think we can make `-O0` a `PackageDistSuspiciousWarn` when it is used in test suites/benchmarks.

**Comment by goldfirere** [comment]
Yes I suppose you're right, @hvr, in that it could be more of a local setup. I wouldn't want clients who want to run the testsuite to pay the cost of optimizing (it's non-negligible) but I guess it's not so terrible.

Does `cabal configure` allow for different options for the testsuite than for the library?

**Comment by hvr** [comment]
> Does cabal configure allow for different options for the testsuite than for the library?

not *yet*, but it's an instance of the more general feature to be able to not only write something like

```
package aeson
    profiling-detail: toplevel-functions
    optimization: 2
    ghc-options: -Wcompat -Wall

package text
    profiling-detail: toplevel-functions
    optimization: False
```

but instead be able to say something akin to (strawman):

```
-- example assumes best-match-wins semantics
target aeson:lib:aeson
   optimization: 2

target aeson:test:*
    optimization: True

target *:test:*
    optimization: False
    ghc-options: -Wall -Wcompat
```

to specify in a more granular way (and with globbing to avoid tedious repetition) where to inject certain flags.

**Comment by ezyang** [comment]
I'm OK with making the warning for `-O0` non-fatal. If you like, we could keep it fatal for libraries.

**Comment by hvr** [comment]
@ezyang so let's make it non-fatal only for tests & benchmarks?

**Comment by simonmichael** [comment]
-O0 is also used in the standard yesod project config when the dev or library-only flags are active, to speed up developer rebuils.

**Comment by nomeata** [comment]
Maybe a corner case, but `ghc-heap-view` requires `-O0` to work properly (for reasons that I don't quite remember – basically I don’t want GHC to mess too much with fragile unsafe code). I disabled `cabal check` on my CI for now…

**Comment by 23Skidoo** [comment]
Fixed by #4967.
